main.tf

```diff
@@ -1 +1,346 @@
+
+  tags = {
+    Environment = "${var.environment}"
+  }
+}
+
+###
+### DATA SOURCES
+###
+
+data "aws_partition" "current" {}
+
+data "aws_caller_identity" "current" {}
+
+data "aws_region" "current" {}
+
+data "aws_ami" "this" {
+  most_recent = "true"
+
+  owners = ["${var.ami_owner}"]
+
+  name_regex = "${local.ami_name_regex}"
+
+  filter {
+    name   = "name"
+    values = ["${local.ami_name_filter}"]
+  }
+}
+
+data "http" "ip" {
+  url = "http://ipv4.icanhazip.com"
+}
+
+data "aws_subnet" "lb" {
+  count = "${length(var.lb_subnet_ids)}"
+
+  id = "${var.lb_subnet_ids[count.index]}"
+}
+
+data "aws_kms_key" "this" {
+  key_id = "${var.kms_key_id}"
+}
+
+data "template_file" "instance_policy" {
+  template = "${file("policies/iam_policy.json")}"
+
+  vars {
+    partition  = "${data.aws_partition.current.partition}"
+    region     = "${data.aws_region.current.name}"
+    account_id = "${data.aws_caller_identity.current.account_id}"
+
+    name           = "${var.name}"
+    environment    = "${var.environment}"
+    key_id         = "${data.aws_kms_key.this.key_id}"
+    dynamodb_table = "${var.dynamodb_table}"
+    bucket_name    = "${local.bucket}"
+    ssm_path       = "${local.ssm_root_path}"
+  }
+}
+
+data "aws_iam_policy_document" "instance_trust_policy" {
+  statement {
+    actions = ["sts:AssumeRole"]
+
+    principals {
+      type        = "Service"
+      identifiers = ["ec2.${var.url_suffix}"]
+    }
+  }
+}
+
+# Manage IAM Roles and Policies
+resource "aws_iam_role" "instance" {
+  name               = "${local.policy_name}"
+  assume_role_policy = "${data.aws_iam_policy_document.instance_trust_policy.json}"
+}
+
+resource "aws_iam_role_policy" "instance" {
+  name_prefix = "${local.policy_name}_"
+  policy      = "${data.template_file.instance_policy.rendered}"
+  role        = "${aws_iam_role.instance.id}"
+}
+
+resource "aws_iam_instance_profile" "instance" {
+  name = "${local.policy_name}"
+  role = "${aws_iam_role.instance.name}"
+}
+
+# Generate a random id for each deployment
+resource "random_string" "this" {
+  length  = 8
+  special = "false"
+}
+
+# Manage bucket contents
+locals {
+  s3_sync_salt = [
+    "aws s3 sync --delete",
+    "${path.module}/salt",
+    "s3://${local.bucket}/${random_string.this.result}/salt",
+  ]
+
+  s3_sync_appscript = [
+    "aws s3 sync --delete",
+    "${path.module}/scripts",
+    "s3://${local.bucket}/${random_string.this.result}/scripts",
+  ]
+
+  s3_rm = [
+    "aws s3 rm --recursive",
+    "s3://${local.bucket}/${random_string.this.result}",
+  ]
+}
+
+resource "null_resource" "sync_s3" {
+  provisioner "local-exec" {
+    command = "${join(" ", local.s3_sync_salt)}"
+  }
+
+  provisioner "local-exec" {
+    command = "${join(" ", local.s3_sync_appscript)}"
+  }
+
+  provisioner "local-exec" {
+    command = "${join(" ", local.s3_rm)}"
+    when    = "destroy"
+  }
+
+  triggers = {
+    s3_sync_salt      = "${join(" ", local.s3_sync_salt)}"
```

Comment:
this was a bit of a quick hack in the cats-or-dogs project... this will trigger updates with the _command_ changes, but not when the _content_ changes. to support syncing updated content, we could probably use a hash of the content. to get a hash of an entire directory, probably need to [create an archive](https://www.terraform.io/docs/providers/archive/d/archive_file.html)